reasoner/resolution/resolver/CompoundResolver.java

```diff
@@ -0,0 +1,148 @@
+                               ConceptManager conceptMgr, boolean resolutionTracing, Actor<ResolutionRecorder> resolutionRecorder) {
+        super(self, name, registry, traversalEngine, conceptMgr, resolutionTracing);
+        this.resolutionRecorder = resolutionRecorder;
+        this.responses = new HashMap<>();
+        this.isInitialised = false;
+    }
+
+    protected abstract void nextAnswer(Request fromUpstream, R responseProducer, int iteration);
+
+    @Override
+    public void receiveRequest(Request fromUpstream, int iteration) {
+        LOG.trace("{}: received Request: {}", name(), fromUpstream);
+        if (!isInitialised) initialiseDownstreamResolvers();
+        if (isTerminated()) return;
+
+        R responses = getOrUpdateResponses(fromUpstream, iteration);
+        if (iteration < responses.iteration()) {
+            // short circuit if the request came from a prior iteration
+            failToUpstream(fromUpstream, iteration);
+        } else {
+            assert iteration == responses.iteration();
+            nextAnswer(fromUpstream, responses, iteration);
+        }
+    }
+
+    @Override
+    protected void receiveFail(Response.Fail fromDownstream, int iteration) {
+        LOG.trace("{}: received Exhausted, with iter {}: {}", name(), iteration, fromDownstream);
+        if (isTerminated()) return;
+
+        Request toDownstream = fromDownstream.sourceRequest();
+        Request fromUpstream = fromUpstream(toDownstream);
+        R responseProducer = responses.get(fromUpstream);
+
+        if (iteration < responseProducer.iteration()) {
+            // short circuit old iteration failed messages back out of the actor model
+            failToUpstream(fromUpstream, iteration);
+            return;
+        }
+        responseProducer.removeDownstreamProducer(fromDownstream.sourceRequest());
+        nextAnswer(fromUpstream, responseProducer, iteration);
+    }
+
+    private R getOrUpdateResponses(Request fromUpstream, int iteration) {
+        if (!responses.containsKey(fromUpstream)) {
+            responses.put(fromUpstream, responsesCreate(fromUpstream, iteration));
+        } else {
+            R responses = this.responses.get(fromUpstream);
+            assert responses.iteration() == iteration ||
+                    responses.iteration() + 1 == iteration;
+
+            if (responses.iteration() + 1 == iteration) {
+                // when the same request for the next iteration the first time, re-initialise required state
+                R responseProducerNextIter = responsesReiterate(fromUpstream, responses, iteration);
+                this.responses.put(fromUpstream, responseProducerNextIter);
+            }
+        }
+        return responses.get(fromUpstream);
+    }
+
+    abstract R responsesCreate(Request fromUpstream, int iteration);
+
+    abstract R responsesReiterate(Request fromUpstream, R priorResponses, int iteration);
+
+    static class Responses {
```

Comment:
Is there a better name than `Responses`? i don't feel like it captures what this class does - it's not just a set of responses.